File: src/services/payments.ts
/**
 * Client for the Neurotech Events API payment endpoints.
 *
 * Live payments are enabled only when `VITE_API_BASE_URL` is configured. The server owns
 * prices, provider calls and confirmation; this module only relays requests and statuses.
 */
import type { PaymentMethod, PaymentStatus } from "../domain/types";
import { ApiError, apiRequest, isApiEnabled, optionalAuthRequest } from "./api";

export const isLivePaymentsEnabled = isApiEnabled;
export const PaymentApiError = ApiError;
export type PaymentApiError = ApiError;

const request = apiRequest;

export type MobileMethod = Extract<PaymentMethod, "mpesa" | "airtel" | "mixx" | "halopesa">;

export interface StartMobilePaymentInput {
  event_slug: string;
  ticket_code: string;
  phone_number: string;
  method: MobileMethod;
  attendee: { full_name: string; email: string; organization?: string; job_title?: string; country?: string };
  client_reference?: string;
}

export interface RemotePaymentEvent {
  from_status: string | null;
  to_status: string;
  note: string | null;
  created_at: string;
}

export interface RemotePayment {
  id: string;
  reference: string;
  status: PaymentStatus;
  amount: number;
  currency: string;
  method: PaymentMethod;
  provider: string | null;
  provider_reference: string | null;
  registration_id: string;
  registration_status: string;
  event_id: string;
  attendee_id: string;
  ticket_type_id: string;
  ticket_number: string;
  event_slug: string;
  event_title: string;
  ticket_name: string;
  attendee_name: string;
  attendee_email: string;
  created_at: string;
  updated_at: string;
  paid_at: string | null;
  events: RemotePaymentEvent[];
}

export interface RemotePaymentPage {
  items: RemotePayment[];
  total: number;
  page: number;
  page_size: number;
}

export interface ProviderTransaction {
  reference?: string;
  id?: string;
  status?: string;
  amount?: { value?: number; currency?: string } | number;
  channel?: { type?: string; provider?: string };
  customer?: { first_name?: string; last_name?: string; name?: string; email?: string; phone?: string };
  settlement?: { gross?: { value?: number }; fees?: { value?: number }; net?: { value?: number } };
  metadata?: Record<string, unknown>;
  created_at?: string;
  completed_at?: string;
  failure_reason?: string | null;
  payment_type?: string;
  external_reference?: string;
}

export interface ProviderTransactionsPage {
  provider: string;
  items: ProviderTransaction[];
  raw: Record<string, unknown>;
}

export interface ProviderBalance {
  provider: string;
  available: number;
  balance: number;
  currency: string;
}

export function startMobilePayment(input: StartMobilePaymentInput): Promise<RemotePayment> {
  // Signed-in buyers are linked to their account so the ticket shows up under "My tickets".
  return optionalAuthRequest<RemotePayment>("/payments/mobile", { method: "POST", body: JSON.stringify(input) });
}

export function fetchPayment(paymentId: string): Promise<RemotePayment> {
  return request<RemotePayment>(`/payments/${encodeURIComponent(paymentId)}`);
}

export function fetchAdminPayments(token: string, params: { status?: string; page?: number; page_size?: number } = {}): Promise<RemotePaymentPage> {
  const query = new URLSearchParams();
  if (params.status) query.set("status", params.status);
  if (params.page) query.set("page", String(params.page));
  if (params.page_size) query.set("page_size", String(params.page_size));
  const suffix = query.toString() ? `?${query}` : "";
  return request<RemotePaymentPage>(`/admin/payments${suffix}`, {}, token);
}

export function fetchProviderTransactions(token: string, params: { limit?: number; page?: number; status?: string } = {}): Promise<ProviderTransactionsPage> {
  const query = new URLSearchParams();
  if (params.limit) query.set("limit", String(params.limit));
  if (params.page) query.set("page", String(params.page));
  if (params.status) query.set("status", params.status);
  const suffix = query.toString() ? `?${query}` : "";
  return request<ProviderTransactionsPage>(`/admin/payments/provider${suffix}`, {}, token);
}

export function fetchProviderBalance(token: string): Promise<ProviderBalance> {
  return request<ProviderBalance>("/admin/payments/balance", {}, token);
}

export function verifyAdminPayment(token: string, paymentId: string): Promise<RemotePayment> {
  return request<RemotePayment>(`/admin/payments/${encodeURIComponent(paymentId)}/verify`, { method: "POST" }, token);
}

export function providerAmount(item: ProviderTransaction): number {
  if (typeof item.amount === "number") return item.amount;
  return item.amount?.value ?? 0;
}

export function providerCustomer(item: ProviderTransaction): string {
  const c = item.customer;
  if (!c) return "—";
  return c.name ?? [c.first_name, c.last_name].filter(Boolean).join(" ") ?? c.email ?? c.phone ?? "—";
}

export interface TicketQuote {
  event_slug: string;
  code: string;
  name: string;
  tier: string;
  currency: string;
  price: number;
  vat_percent: number | string;
  vat: number;
  total: number;
  active: boolean;
  capacity: number;
  sold: number;
  available: number;
  payable_online: boolean;
}

/** The exact amount the server will charge for a ticket, VAT included. */
export function fetchTicketQuote(eventSlug: string, ticketCode: string): Promise<TicketQuote> {
  return request<TicketQuote>(`/events/${encodeURIComponent(eventSlug)}/tickets/${encodeURIComponent(ticketCode)}/quote`);
}

export interface CatalogueSyncResult {
  events_upserted: number;
  ticket_types_upserted: number;
  ticket_types_deactivated: number;
}

export function syncCatalogue(token: string, payload: unknown): Promise<CatalogueSyncResult> {
  return request<CatalogueSyncResult>("/admin/catalogue", { method: "PUT", body: JSON.stringify(payload) }, token);
}

File: backend/app/schemas/payments.py
import re
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

MobileMethod = Literal["mpesa", "airtel", "mixx", "halopesa"]
_EMAIL = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


class AttendeeInput(BaseModel):
    full_name: str = Field(min_length=2, max_length=200)
    email: str = Field(max_length=320)
    organization: str | None = Field(default=None, max_length=200)
    job_title: str | None = Field(default=None, max_length=200)
    country: str | None = Field(default=None, max_length=120)

    @field_validator("email")
    @classmethod
    def _email(cls, value: str) -> str:
        value = value.strip().lower()
        if not _EMAIL.match(value):
            raise ValueError("must be a valid email address")
        return value


class MobilePaymentRequest(BaseModel):
    event_slug: str = Field(min_length=1, max_length=160)
    ticket_code: str = Field(min_length=1, max_length=60)
    phone_number: str = Field(min_length=9, max_length=20)
    method: MobileMethod = "mpesa"
    attendee: AttendeeInput
    # Opaque reference from the calling client (e.g. its local registration id) for correlation.
    client_reference: str | None = Field(default=None, max_length=60)

    @field_validator("phone_number")
    @classmethod
    def _phone(cls, value: str) -> str:
        digits = "".join(ch for ch in value if ch.isdigit())
        if digits.startswith("0") and len(digits) == 10:
            digits = "255" + digits[1:]
        if not (len(digits) == 12 and digits.startswith("255")):
            raise ValueError("must be a Tanzanian mobile number such as 255712345678 or 0712345678")
        return digits


class FreeRegistrationRequest(BaseModel):
    """Register for a ticket that costs nothing; the server confirms the total is zero."""

    event_slug: str = Field(min_length=1, max_length=160)
    ticket_code: str = Field(min_length=1, max_length=60)
    attendee: AttendeeInput
    phone_number: str | None = Field(default=None, max_length=20)


class PaymentEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    from_status: str | None
    to_status: str
    note: str | None
    created_at: datetime


class PaymentOut(BaseModel):
    id: uuid.UUID
    reference: str
    status: str
    amount: int
    currency: str
    method: str
    provider: str | None
    provider_reference: str | None
    registration_id: uuid.UUID
    registration_status: str
    event_id: uuid.UUID
    attendee_id: uuid.UUID
    ticket_type_id: uuid.UUID
    ticket_number: str
    event_slug: str
    event_title: str
    ticket_name: str
    attendee_name: str
    attendee_email: str
    created_at: datetime
    updated_at: datetime
    paid_at: datetime | None
    events: list[PaymentEventOut]


class PaymentPageOut(BaseModel):
    items: list[PaymentOut]
    total: int
    page: int
    page_size: int


class ProviderTransactionsOut(BaseModel):
    provider: str
    items: list[dict[str, Any]]
    raw: dict[str, Any]


class BalanceOut(BaseModel):
    provider: str
    available: int
    balance: int
    currency: str
